From 5f3e7ba1cff32038845b07ea616b8dc6e8bcd1cb Mon Sep 17 00:00:00 2001
From: NaKyle <nakyle.wright@gmail.com>
Date: Mon, 23 Dec 2024 20:24:49 -0800
Subject: [PATCH 099/188] cec: cros_ec buffers messages, read all available
 rather than just 1.

(cherry picked from commit 2ae417bcf7c0c0818cbdfe053e836790440e8129)
Signed-off-by: NaKyle <nakyle.wright@gmail.com>
---
 .../media/cec/platform/cros-ec/cros-ec-cec.c  | 25 +++++++++++++------
 1 file changed, 18 insertions(+), 7 deletions(-)

diff --git a/drivers/media/cec/platform/cros-ec/cros-ec-cec.c b/drivers/media/cec/platform/cros-ec/cros-ec-cec.c
index 6c8b50afb..30538262f 100644
--- a/drivers/media/cec/platform/cros-ec/cros-ec-cec.c
+++ b/drivers/media/cec/platform/cros-ec/cros-ec-cec.c
@@ -101,17 +101,28 @@ static void cros_ec_cec_read_message(struct cros_ec_cec_port *port)
 		.port = port->port_num,
 	};
 	struct ec_response_cec_read response;
-	int ret;
+	int ret = 0;
+	int count = 0;
 
-	ret = cros_ec_cmd(cros_ec, 0, EC_CMD_CEC_READ_MSG, &params,
-			  sizeof(params), &response, sizeof(response));
-	if (ret < 0) {
+	/* NOTE read as many messages as we can, return value from cros_ec_cmd
+	 *      here should either equal to the bytes transferred or a negative
+	 *      error code, so as long as we get positive integers we are
+	 *      reading messages!
+	 */
+	do {
+		ret = cros_ec_cmd(cros_ec, 0, EC_CMD_CEC_READ_MSG, &params,
+				  sizeof(params), &response, sizeof(response));
+		if (ret > 0) {
+			++count;
+			cros_ec_cec_received_message(port, response.msg, response.msg_len);
+		}
+	}
+	while (ret > 0);
+
+	if (ret < 0 && count == 0) {
 		dev_err(cros_ec->dev,
 			"error reading CEC message on EC: %d\n", ret);
-		return;
 	}
-
-	cros_ec_cec_received_message(port, response.msg, response.msg_len);
 }
 
 static void handle_cec_event(struct cros_ec_cec *cros_ec_cec)
-- 
2.55.0

